### PR TITLE
#107 Extract Enceladus utils into commons and reorganize implicits 

### LINKED ISSUE #107: Extract Enceladus utils into commons
### Background
[Enceladus](https://github.com/AbsaOSS/enceladus) has some functionality that can be useful outside.

### Goal
Extract part of `utils` module of Enceladus (that is not tight to it) to this repo.

### PR DESCRIPTION
**Pull request by jakipatryk**
#### Moved from Enceladus `utils` module:
- `implicits.OptionImplicits` -> `lang.extensions.OptionExtension`
- `implicits.StringImplicits` -> `lang.extensions.StringExtension`
- `general.Algorithms` -> `lang.extensions.SeqExtension`
- `general.ClassLoaderUtils.loadSingletonClassOfType` -> `reflect.ReflectionUtils.objectForNameWithDescriptiveException`
- `fs.LocalFsUtils` -> partially (without `getLocalTemporaryDirectory`, as this functionality is already in `io.TempDirectory`) `io.LocalFileSystemUtils`
- `[Menas/RestApi/Uris]ConnectionStringParser` -> `parsers.UrisConnectionStringParser`

#### On the fly [changes between Enceladus and commons]:
- added tests for `lang.extensions.OptionExtension`
- `nonEmpyOrElse` -> `nonEmptyOrElse` (fixed typo, `lang.extensions.StringExtension`)
- deduplicated code in `lang.extensions.SeqExtension` (in Enceladus these are 3 function with full body, here in commons there is one function with full body, and the other two are using the first one)
- renamed functions in `lang.extensions.SeqExtension` (in Enceladus `stableGroupBy...`, here in commons `groupConsecutiveBy...`)
- in `UrisConnectionStringParser.parse` removed deleting `/api` suffix as this is Enceladus-specific
- added tests for `reflect.ReflectionUtils.objectForNameWithDescriptiveException`
- changed implicit classed into implicit value classes

#### Restructured implicits in commons
Before this PR, there were two implicit files in `lang` - `CollectionImplicits` and `OptionImplicits`. However, these two have somewhat inconsistent convention of what should go in them. To solve that, I created subpackage `lang.extensions` and moved methods from previous implicit files there following convention:

- file name = object name = `TypeExtension` (extending capabilities of a given type)
- implicit class name = `TypeOps` (the same `Type` as in object name)

`CollectionImplicits` and `OptionImplicits` stay in the library for backwards compatibility, but are deprecated.

Also, implicits moved from Enceladus follow the new convention.

Closes #107

### CODE REVIEW

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala [hunk 1]
```diff
@@ -81,16 +82,47 @@ object ReflectionUtils {
...
+  def objectForNameWithDescriptiveException[T <: AnyRef : ClassTag : TypeTag](name: String): T = {
+    Try(mirror.staticModule(name)) match {
+      case Success(module) =>
+        val reflectiveMirror = mirror.reflectModule(module)
+        Try(reflectiveMirror.instance) match {
+          case Success(instance) =>
+            instance match {
+              case singleton: T => singleton
+              case _ =>
+                throw new IllegalArgumentException(s"Class '$name' is not an instance of '${typeOf[T]}'")
```
Comment (wajda) [inline src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:116]: Why is `IllegalArgumentException` more descriptive than a reflection exception thrown from a reflection operation?

File: src/main/scala/za/co/absa/commons/parsers/UrisConnectionStringParser.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.commons.parsers
+
+object UrisConnectionStringParser {
```
Comment (wajda) [inline src/main/scala/za/co/absa/commons/parsers/UrisConnectionStringParser.scala:18]: What is the use-case for this utility? Can it be considered generic enough to belong commons?

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala (hunk 1 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:116]: The descriptiveness in the name of this function comes mostly from the case when provided `name` is not a singleton. In that case, `objectForName` throws `java.lang.ClassNotFoundException: <name>`, and `objectForNameWithDescriptiveException` throws `IllegalArgumentException: Class '<name>' is not a singleton`.

However, the more I think about it, the more I'm convinced that this function is not needed, and I'll probably remove it from this PR.

File: src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala [hunk 3]
```diff
@@ -90,36 +90,11 @@ object ReflectionUtils {
...
-          case Failure(exception) =>
-            throw new IllegalArgumentException(s"Class '$name' is not a singleton", exception)
-        }
-      case Failure(exception) =>
-        throw new IllegalArgumentException(s"Class '$name' could not be found", exception)
+    try {
+      val moduleSymbol = mirror.staticModule(name)
+      mirror.reflectModule(moduleSymbol).instance.asInstanceOf[T]
+    } catch {
+      case _: ClassNotFoundException => throw new ClassNotFoundException(s"Class '$name' is not a singleton")
```
Comment (wajda) [inline src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala:97]:
```suggestion
      case e: ClassNotFoundException => throw new ClassNotFoundException(s"Class '$name' is not a singleton", e)
```
we probably don't want to lose the original stack trace

File: src/main/scala/za/co/absa/commons/parsers/UrisConnectionStringParser.scala (hunk 2 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/parsers/UrisConnectionStringParser.scala]: I decided to delete it as indeed it seems to be too Enceladus-specific and similar functionality can be achieved with less effort in other/future projects.

**Review by wajda — APPROVED** [review]
LGTM

File: src/main/scala/za/co/absa/commons/lang/extensions/OptionExtension.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  implicit class OptionOps[T](val option: Option[T]) extends AnyVal {
+
+    /**
+     * Converts Option to Try such that Some(x) becomes Success(x), and None becomes Failure(exception),
+     * where exception is provided as parameter.
+     *
+     * @param failure an Exception to use in Failure in case Option is None
+     * @return Try converted from Option
+     */
+    def toTry(failure: Exception): Try[T] =
```
Comment (benedeki) [inline src/main/scala/za/co/absa/commons/lang/extensions/OptionExtension.scala:32]:
```suggestion
    def toTry(failure: =>Exception): Try[T] =
```
No need to instantiate the exception if not needed

File: src/main/scala/za/co/absa/commons/lang/extensions/OptionExtension.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      option.fold[Try[T]](Failure(failure))(Success(_))
```
Comment (benedeki) [inline src/main/scala/za/co/absa/commons/lang/extensions/OptionExtension.scala:33]: But considering the code uses `fold` with initial value instantiating the `Failure`, perhaps this code would be better:
```suggestion
    def toTry(failure: =>Exception): Try[T] =
      option.map(Success(_)).getOrElse(Failure(failure))
```

File: src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+     * "abc" / "/123" -> "abc/123"
+     * "abc/" / "/123" -> "abc/123",
+     * but:
+     * "file:///" / "path" -> "file:///path",
+     * }}}
+     *
+     * @param another the second string we are appending after the `/` separator
+     * @return this/another (this has stripped trailing / if present, another has leading / stripped if present)
+     */
+    def /(another: String): String = { // scalastyle:ignore method.name
```
Comment (benedeki) [inline src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala:141]: Is this something worth to have in _commons_?

File: src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+package za.co.absa.commons.lang.extensions
+
+object TraversableExtension {
+
+  implicit class TraversableOps[A <: Traversable[_]](val xs: A) extends AnyVal {
+
+    def asOption: Option[A] = if (xs.isEmpty) None else Some(xs)
```
Comment (benedeki) [inline src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala:23]: Isn't this covered in [`NonOptionExtension`]([NonOptionExtension.scala](https://github.com/AbsaOSkjk,.S/commons/pull/108/files#diff-154e0e72b8f99fdf0051419d6dec1408e403c782bf6b5fbff3943ef3565cab7b))?

Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala]: No this is different. The one in `NonOptionExtension` would return `Some(List())` and the one here `None` if called on empty list. I committed an update to above method that allows being called on nulls + added a description of the method.

File: src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala (hunk 6 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala:141]: Good question. To me, this is the only natural usage of `/` operator on strings, as I'm not aware of any natural or at least somewhat commonly used [field](https://en.wikipedia.org/wiki/Field_(mathematics)) structure on strings. So in my opinion it is worth to have it here.

File: src/main/scala/za/co/absa/commons/lang/extensions/OptionExtension.scala (hunk 5 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/extensions/OptionExtension.scala]: Agree, changed.

File: src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala (hunk 7 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala]: it would also be good to cover added boundary conditions with tests

Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala]: Added tests for TraversableExtension

File: src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala (hunk 6 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala:141]: Good spotted. I also think it's not quite intuitive. By analogy with the standard `*` operator on `String` I would expect `/` would do the opposite, e.i.:
```scala
"abc" * 3 == "abcabcabc"
"abcabcabc" / 3 == "abc"
```
(not telling it's something practically useful thought, but such semantics would just be less of surprise to me I guess)

In your case `/` is used for a specific case of string concatenation, something like `+` but smarter :)
I would say you can accomplish almost the same by simply using `Path.of(a, b).toString` without involving any additional commons function call. That would cover all but the last case (the one with `file:///`)

```scala
     val myPath = Path.of("abc/", "/123").toString // self explanatory
     // versus
     val myPath = "abc/" / "/123" // kinda magic
```
The last `file:///` case, I'm not sure how exactly you use it in Enceladus, but I would guess the string `file:///` is likely only exist as a literal in the code. If so than a simple string concat would do the same thing.

Also, for string concatenation it's recommended to prefer string interpolation over `+` or other ways.

Comment (dk1844) [reply src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala:141]: > Is this something worth to have in _commons_?

Of course, I am author-biased, but I find it quite useful for cases where user's input on path is expected and concatenation of paths (specifically prefixes, relative path and file names) is happening. Without this, it is tiresome to otherwise check trailing/leading slashes when doing e.g.
```scala
rootPath / relativePath / fileName
```
where the `rootPath` for example may be given both as `/path/to/here` or `/path/to/here/`, and analogically for the rest.

With @wajda's examples of
```scala
"abc" * 3 == "abcabcabc"
"abcabcabc" / 3 == "abc"
```
I agree that this is (`*`) / could be (`/`) the expected behavior, but this is specialized to `Int` params. With a `String` param, the expectation, I believe can be different.

My vote on this matter would be, since you apparently find it not generic enough to stay in generic `StringOps` implicit, which I can see the motivation for, to move this into a different implicit wrapper, perhaps named `PathJoin`, so this can still be (re)used easily, but within a proper context.

File: src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala (hunk 7 above)
Comment (benedeki) [reply src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala]: Thanks for the explanation, of course it's different. Cool with the added comment.

File: src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala (hunk 6 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala:141]: > move this into a different implicit wrapper
that would be fine IMO

Although why is `Path.of(rootPath, relativePath, fileName)` not enough?

Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/extensions/StringExtension.scala:141]: As far as I know, Path.of is platform-specific (different separator on Windows and rest of the world), wheras `/` would be platform-agnostic.

I agree to move it to a separate implicit class, but I'm gonna name it `StringConcatenationOps`, as the `/` functionality can be used also outside paths context. What do you think? 

Then the convention for implicit extensions in the library would be as I have written in the description of the PR + if some functionality is not general enough for a given type put it in `TypeExtension.Type<Usecase>Ops`.

File: src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala (hunk 7 above)
Comment (dk1844) [reply src/main/scala/za/co/absa/commons/lang/extensions/TraversableExtension.scala]: An good idea with the null check 👍

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala [hunk 8]
```diff
@@ -21,6 +21,7 @@ import scala.collection.mutable
 import scala.language.higherKinds
 import scala.reflect.ClassTag
 
+@deprecated("Use type-specific ...Extension instead")
```
Comment (dk1844) [inline src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:24]: Shouldn't this be accompanied by the current version to singify which version it was deprecated from? (multiple occurrences)

File: src/main/scala/za/co/absa/commons/lang/extensions/AnyExtension.scala [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Applies `applyFn` to the value with `maybeArg` as second argument in case `maybeArg` is not None,
+     * otherwise does nothing and returns the same value as was called on.
+     */
+    def optionally[B](applyFn: (A, B) => A, maybeArg: Option[B]): A = maybeArg.map(applyFn(a, _)).getOrElse(a)
+
+    /**
+     * The same as [[optionally]], but with signature that is more type inference friendly.
+     */
+    def having[B](maybeArg: Option[B])(applyFn: (A, B) => A): A = optionally(applyFn, maybeArg)
```
Comment (dk1844) [inline src/main/scala/za/co/absa/commons/lang/extensions/AnyExtension.scala:32]: perhaps at least a simple single test case can exists for this in AnyOpsSpec, too?
(also given the fact that this is a README-marketed feature)

File: src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * Reads a local file fully and returns its content.
+   *
+   * @param path A path to a file.
+   * @return The file's content.
+   */
+  def readLocalFile(path: String): String = {
+    Files.readAllLines(Paths.get(path), StandardCharsets.UTF_8).toArray.mkString("\n")
+  }
```
Comment (Zejnilovic) [inline src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala:40]: Should the charset be a parameter with a default value?

**Review by dk1844 — DISMISSED** [review]
Nicely done (LGTM, just read the code).
I also appreciate the scala doc enhancements here and there

**Review by Zejnilovic — APPROVED** [review]
Code reviewed

File: src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala (hunk 10 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala:40]: I would suggest `org.apache.commons.io.FileUtils.readFileToString()` from Apache Commons IO as an existing alternative method that does exactly the same thing but directly, without pointless intermediate splitting into lines followed by concatenating everything back in one piece.

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala (hunk 8 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:24]: Here is the thing - what version is the new version gonna be? I've heard we do semantic versioning in this library, but as far as I see, we pretty much always bump patch version (x.x.patch). So - is it gonna be 1.0.8, or 1.1.0?

File: src/main/scala/za/co/absa/commons/lang/extensions/AnyExtension.scala (hunk 9 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/extensions/AnyExtension.scala:32]: Hmm I pretty much never test alias functions (as they do not provide new runtime features), especially when they are marked as aliases in docstring, that's why I didn't create such.

File: src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala (hunk 10 above)
Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala]: Added charset parameter and I agree that `org.apache.commons.io.FileUtils.readFileToString()` can be used as an alternative. But given that I didn't know about this function before your comment, maybe we should keep `readLocalFile` in our commons? Here we have a smaller set of functionalities that in a way is easier to discover.

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala (hunk 8 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:24]: **1.1.0**

SemVer 2.0 spec makes it clear:
> Bug fixes not affecting the API increment the patch version, backwards compatible API additions/changes increment the minor version, and backwards incompatible API changes increment the major version.

Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:24]: Yeah I know, I was just confused because so far even on new features patch version was bumped, not minor (which should be).

File: src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala (hunk 10 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala]: Maybe with `@deprecated` annotation then? I think it would be useful for a person that has found this function to know that there is a better alternative?

Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/io/LocalFileSystemUtils.scala]: Makes sense, will add that.

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala (hunk 8 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:24]: Yeah, the automated release pipeline always increment patch version. So unless it's not explicitly intercepted on the PR level it's out there automatically. We need to be more careful about it. :+1:

Comment (jakipatryk) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala]: Changed version to 1.1.0-SNAPSHOT in the pom.xml. Is there anything else that needs to be done to make the plugin working as expected for this PR?

**Review by wajda — APPROVED** [review]
LGTM

File: src/main/scala/za/co/absa/commons/lang/extensions/AnyExtension.scala (hunk 9 above)
Comment (dk1844) [reply src/main/scala/za/co/absa/commons/lang/extensions/AnyExtension.scala:32]: I understand the reasoning, I never meant to include a full set of tests comparable to one for `optionally`. 
Not pushing, just trying to explain.

My idea was merely to provide at least one test case that serves as evidence that `having` works as an alias of `optionally`. This test would simply fail should the implementation change (or if the method gets removed), given that `having` is about to become a part of the public API on this package.

Your call, though, I am fine if left as-is, too.

**Review by dk1844 — APPROVED** [review]
